### PR TITLE
WIP: Openresty 119 fixes

### PR DESCRIPTION
**Pull request by csmithATsquiz**

### CODE REVIEW

File: t/01-unit/cache_key.t [hunk 1]
```diff
@@ -330,14 +330,17 @@ location /t {
         local vary_key = generate_vary_key({"Foo", "X-Test"}, callback, nil)
         log(vary_key)
         assert(called_flag == true, "Callback is called - multivalue spec")
-        assert(vary_key == "foo:bar:x-test:value", "Vary spec not modified with noop function - multivalue spec")
+        assert(string.find(vary_key, "foo:bar"), "Vary spec not modified with noop function - multivalue spec")
```
Comment (hamishforbes) [inline t/01-unit/cache_key.t:333]: This wouldn't catch an addition to the `vary_key`, might be worth just checking the overall length of the string too?

Comment (csmithATsquiz) [reply t/01-unit/cache_key.t:333]: Thanks, I've added that now.

Comment (csmithATsquiz) [reply t/01-unit/cache_key.t:333]: Would these sort of checks be better done as something like this:
```
        assert(string.find(vary_key, "foo:bar"), "Vary spec not modified with noop function - multivalue spec checking for foo-bar")
        vary_key = string.gsub(vary_key, "foo:bar", "")
        assert(string.find(vary_key, "x-test:value"), "Vary spec not modified with noop function - multivalue spec checking for x-test")
        vary_key = string.gsub(vary_key, "x?-test:value", "")
        assert(vary_key == ":", "Vary spec not modified with noop function - multivalue spec looking for :")
```

Comment (hamishforbes) [reply t/01-unit/cache_key.t:333]: There's a risk of being 'too clever' in these tests though, keeping it as simple as possible means any bizarre unexpected values would get caught.
You could parse the string, split it by `:` and check that all the expected values are there and no unexpected values are. 
🤷‍♂️  probably fine as it is now

**Comment by pintsized** [comment]
FYI I've pushed versioned docker images for the test runner, based on the last three major OpenResty images, so that it's possible to run tests against different versions instead of just an out-of-date `latest`. Note that from `1.17.*` onward the OpenResty images are from Centos 8, which took a bit of messing around (and may give different results to your upgraded Centos 7, so lets see).

I bumped the `latest` tag too, so next time you push to this branch it should run against `1.19.3.1` on travis.

https://hub.docker.com/repository/docker/ledgetech/test-runner

**Comment by pintsized** [comment]
@hamishforbes reckon you can check my changes to the vary code? Some parts of the tests seem to have been passing by luck more than design I think, but the main thing was ensuring the order of the vary key is deterministic. Which means a bit more hot path work, but presumably better HIT rates since the vary key is normalised?

File: t/02-integration/vary.t [hunk 2]
```diff
@@ -445,7 +450,11 @@ location /vary8_prx {
 
 location /vary {
     content_by_lua_block {
-        local incr = ngx.shared.ledge_test:incr("test8", 1, 0)
+        local incr = ngx.shared.ledge_test:incr("test8", 1)
+        if not incr then
+            incr = ngx.shared.ledge_test:incr("test8", 1, 0)
```
Comment (hamishforbes) [inline t/02-integration/vary.t:455]: What's this change for? Why manually check for a failing `incr` without an init argument and then call it again with an init argument? Seems like the behaviour should be identical but this has 3 extra lines?

**Comment by hamishforbes** [comment]
AFAIK given the same LuaJIT build a hash table with the same keys and values will always have the same order with `pairs()`
So because the vary key was being built up from a hash table of `{ header_key = header_val }` it didn't matter in which order the headers were added to that table, they'd always be the same for the same LuaJIT build.

I think your change is *more* correct as it will obviously be the same order across builds, platforms etc but of course with a performance penalty.
I suspect it would make little to no difference in practice though, other than possibly when upgrading

File: t/02-integration/vary.t (hunk 2 above)
Comment (pintsized) [reply t/02-integration/vary.t:455]: Ah yes, I'll revert that. I was getting odd results with the counters and glancing at the docs I (incorrectly) thought it was always re-initialising. I meant to take it out again...

**Comment by pintsized** [comment]
> AFAIK given the same LuaJIT build a hash table with the same keys and values will always have the same order with pairs()

That appears to be what's different now though. We're getting different ordering from one run to the next, not just between builds. And since Lua makes no guarantees by specification, I guess this is what we should be doing.

**Comment by hamishforbes** [comment]
Ah ok, yep I guess it has to sort now then 👍

**Pull request merged**